### PR TITLE
[SHELL32] Handle HotKey and ShowCmd on the shortcut property page

### PR DESCRIPTION
**Pull request by whindsaks**
Read and write the HotKey and ShowCmd settings in the .lnk header.

### CODE REVIEW

File: dll/win32/shell32/CShellLink.cpp [hunk 1]
```diff
@@ -2802,6 +2802,26 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+    /* Run */
+    const WORD runstrings[] = { IDS_SHORTCUT_RUN_NORMAL, IDS_SHORTCUT_RUN_MIN, IDS_SHORTCUT_RUN_MAX };
+    const BYTE runshowcmd[] = { SW_SHOWNORMAL, SW_SHOWMINNOACTIVE, SW_SHOWMAXIMIZED };
+    HWND hRun = GetDlgItem(hwndDlg, IDC_SHORTCUT_RUN_COMBO);
+    for (UINT i = 0; i < 3; ++i)
+    {
+        WCHAR buf[200];
+        if (!LoadString(shell32_hInstance, runstrings[i], buf, _countof(buf))) break;
+        int idx = SendMessage(hRun, CB_ADDSTRING, 0, (LPARAM)buf);
+        if (idx < 0) continue;
```
Comment (binarymaster) [inline dll/win32/shell32/CShellLink.cpp:2817]:
```suggestion
    for (UINT i = 0; i < _countof(runstrings); ++i)
    {
        WCHAR buf[MAX_PATH];
        if (!LoadString(shell32_hInstance, runstrings[i], buf, _countof(buf)))
            break;

        int idx = SendMessage(hRun, CB_ADDSTRING, 0, (LPARAM)buf);
        if (idx < 0)
            continue;
```

File: dll/win32/shell32/CShellLink.cpp [hunk 2]
```diff
@@ -2923,6 +2946,15 @@ LRESULT CShellLink::OnNotify(HWND hwndDlg, int idFrom, LPNMHDR pnmhdr)
 
         HeapFree(GetProcessHeap(), 0, unquoted);
 
+
+        m_Header.wHotKey = SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_KEY_HOTKEY, HKM_GETHOTKEY, 0, 0);
+
+        int idx = (int)SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_RUN_COMBO, CB_GETCURSEL, 0, 0);
+        if (idx != CB_ERR)
```
Comment (learn-more) [inline dll/win32/shell32/CShellLink.cpp:2953]: Is it possible to get `CB_ERR` here?

File: dll/win32/shell32/CShellLink.cpp [hunk 3]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+    {
+        WCHAR buf[MAX_PATH];
+        if (!LoadString(shell32_hInstance, runstrings[i], buf, _countof(buf)))
+            break;
+
+        int idx = SendMessage(hRun, CB_ADDSTRING, 0, (LPARAM)buf);
+        if (idx < 0)
+            continue;
+        SendMessage(hRun, CB_SETITEMDATA, idx, runshowcmd[i]);
+        if (!i || m_Header.nShowCommand == runshowcmd[i])
```
Comment (learn-more) [inline dll/win32/shell32/CShellLink.cpp:2822]: Is there ever a case where `m_Header.nShowCommand == runshowcmd[i]` does not yield a selection?

Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp:2822]: If the .lnk file is corrupted, that is why it sets it to normal just in case (the `!i` part). Windows only allows these 3 values in a shortcut.

The [MS specification ](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-shllink/c3376b21-0931-45e4-b2fc-a48ac0e60d15) spells this out: 

> This value SHOULD be one of the following.
> ...
> All other values MUST be treated as SW_SHOWNORMAL.

File: dll/win32/shell32/CShellLink.cpp (hunk 2 above)
Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp:2953]: In the unlikely case that something in `OnInitDialog` failed. `LoadString` being the prime candidate but also `CB_ADDSTRING`.

Do you feel it is too defensive?

Comment (learn-more) [reply dll/win32/shell32/CShellLink.cpp:2953]: Mostly curiosity.

**Review by learn-more — APPROVED** [review]

**Review by binarymaster — APPROVED** [review]

File: dll/win32/shell32/CShellLink.cpp [hunk 4]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+    for (UINT i = 0; i < _countof(runstrings); ++i)
+    {
+        WCHAR buf[MAX_PATH];
+        if (!LoadString(shell32_hInstance, runstrings[i], buf, _countof(buf)))
+            break;
+
+        int idx = SendMessage(hRun, CB_ADDSTRING, 0, (LPARAM)buf);
+        if (idx < 0)
+            continue;
+        SendMessage(hRun, CB_SETITEMDATA, idx, runshowcmd[i]);
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2821]:
```suggestion
        SendMessageW(hRun, CB_SETITEMDATA, idx, runshowcmd[i]);
```

File: dll/win32/shell32/CShellLink.cpp [hunk 5]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+    const WORD runstrings[] = { IDS_SHORTCUT_RUN_NORMAL, IDS_SHORTCUT_RUN_MIN, IDS_SHORTCUT_RUN_MAX };
+    const BYTE runshowcmd[] = { SW_SHOWNORMAL, SW_SHOWMINNOACTIVE, SW_SHOWMAXIMIZED };
+    HWND hRun = GetDlgItem(hwndDlg, IDC_SHORTCUT_RUN_COMBO);
+    for (UINT i = 0; i < _countof(runstrings); ++i)
+    {
+        WCHAR buf[MAX_PATH];
+        if (!LoadString(shell32_hInstance, runstrings[i], buf, _countof(buf)))
+            break;
+
+        int idx = SendMessage(hRun, CB_ADDSTRING, 0, (LPARAM)buf);
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2818]:
```suggestion
        int idx = SendMessageW(hRun, CB_ADDSTRING, 0, (LPARAM)buf);
```

File: dll/win32/shell32/CShellLink.cpp [hunk 6]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+    SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_KEY_HOTKEY, HKM_SETHOTKEY, m_Header.wHotKey, 0);
+
+    /* Run */
+    const WORD runstrings[] = { IDS_SHORTCUT_RUN_NORMAL, IDS_SHORTCUT_RUN_MIN, IDS_SHORTCUT_RUN_MAX };
+    const BYTE runshowcmd[] = { SW_SHOWNORMAL, SW_SHOWMINNOACTIVE, SW_SHOWMAXIMIZED };
+    HWND hRun = GetDlgItem(hwndDlg, IDC_SHORTCUT_RUN_COMBO);
+    for (UINT i = 0; i < _countof(runstrings); ++i)
+    {
+        WCHAR buf[MAX_PATH];
+        if (!LoadString(shell32_hInstance, runstrings[i], buf, _countof(buf)))
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2815]:
```suggestion
        if (!LoadStringW(shell32_hInstance, runstrings[i], buf, _countof(buf)))
```

File: dll/win32/shell32/CShellLink.cpp [hunk 7]
```diff
@@ -2923,6 +2946,15 @@ LRESULT CShellLink::OnNotify(HWND hwndDlg, int idFrom, LPNMHDR pnmhdr)
 
         HeapFree(GetProcessHeap(), 0, unquoted);
 
+
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2949]:
```suggestion
```

File: dll/win32/shell32/lang/en-US.rc [hunk 8]
```diff
@@ -191,7 +191,7 @@ BEGIN
     LTEXT "Shortcut &key:", IDC_SHORTCUT_KEY, 8, 117, 68, 10
     CONTROL "", IDC_SHORTCUT_KEY_HOTKEY, "msctls_hotkey32", WS_VISIBLE | WS_CHILD | WS_BORDER | WS_TABSTOP, 79, 115, 150, 14
     LTEXT "&Run:", IDC_SHORTCUT_RUN, 8, 136, 68, 10
-    COMBOBOX IDC_SHORTCUT_RUN_COMBO, 79, 134, 150, 14, CBS_DROPDOWNLIST | WS_CHILD | WS_VISIBLE | WS_VSCROLL | WS_TABSTOP
+    COMBOBOX IDC_SHORTCUT_RUN_COMBO, 79, 134, 150, 54, CBS_DROPDOWNLIST | WS_CHILD | WS_VISIBLE | WS_VSCROLL | WS_TABSTOP
```
Comment (HBelusca) [inline dll/win32/shell32/lang/en-US.rc:194]: Why is this necessary? Is this because otherwise we have to scroll in the list to "see" the three choices?

File: dll/win32/shell32/CShellLink.cpp [hunk 9]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+        if (!LoadString(shell32_hInstance, runstrings[i], buf, _countof(buf)))
+            break;
+
+        int idx = SendMessage(hRun, CB_ADDSTRING, 0, (LPARAM)buf);
+        if (idx < 0)
+            continue;
+        SendMessage(hRun, CB_SETITEMDATA, idx, runshowcmd[i]);
+        if (!i || m_Header.nShowCommand == runshowcmd[i])
+        {
+            SendMessage(hRun, CB_SETCURSEL, idx, 0);
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2824]:
```suggestion
            SendMessageW(hRun, CB_SETCURSEL, idx, 0);
```

File: dll/win32/shell32/CShellLink.cpp [hunk 10]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
     if (m_sDescription)
         SetDlgItemTextW(hwndDlg, IDC_SHORTCUT_COMMENT_EDIT, m_sDescription);
 
+    /* Hot key */
+    SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_KEY_HOTKEY, HKM_SETHOTKEY, m_Header.wHotKey, 0);
+
+    /* Run */
+    const WORD runstrings[] = { IDS_SHORTCUT_RUN_NORMAL, IDS_SHORTCUT_RUN_MIN, IDS_SHORTCUT_RUN_MAX };
+    const BYTE runshowcmd[] = { SW_SHOWNORMAL, SW_SHOWMINNOACTIVE, SW_SHOWMAXIMIZED };
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2810]:
```suggestion
    const DWORD runshowcmd[] = { SW_SHOWNORMAL, SW_SHOWMINNOACTIVE, SW_SHOWMAXIMIZED };
```

File: dll/win32/shell32/CShellLink.cpp [hunk 11]
```diff
@@ -2923,6 +2946,15 @@ LRESULT CShellLink::OnNotify(HWND hwndDlg, int idFrom, LPNMHDR pnmhdr)
 
         HeapFree(GetProcessHeap(), 0, unquoted);
 
+
+        m_Header.wHotKey = SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_KEY_HOTKEY, HKM_GETHOTKEY, 0, 0);
```
Comment (HBelusca) [inline dll/win32/shell32/CShellLink.cpp:2950]:
```suggestion
        m_Header.wHotKey = (WORD)SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_KEY_HOTKEY, HKM_GETHOTKEY, 0, 0);
```

**Review by HBelusca — COMMENTED** [review]
Minor stuff, but looks good otherwise :)

File: dll/win32/shell32/lang/en-US.rc (hunk 8 above)
Comment (whindsaks) [reply dll/win32/shell32/lang/en-US.rc:194]: It is the classic combo bug. The height of the control specifies the height of the drop list, the control itself forces the size based on the font no matter what you do (And ComboBoxEx tries and fails to work around this). The bug is only visible with ComCtl v5 on Windows IIRC.

File: dll/win32/shell32/CShellLink.cpp (hunk 10 above)
Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp:2810]: Why? These values are never larger than a byte.

File: dll/win32/shell32/CShellLink.cpp (hunk 4 above)
Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp:2821]: Why? There is no string conversion here.

Comment (HBelusca) [reply dll/win32/shell32/CShellLink.cpp:2821]: For consistency. Why would you here rely on the SendMessage ANSI/UNICODE macro and not just call directly the actual underlying function? (since everywhere else the explicit unicode function is always used)

File: dll/win32/shell32/CShellLink.cpp (hunk 10 above)
Comment (HBelusca) [reply dll/win32/shell32/CShellLink.cpp:2810]: Those values are stored in the combobox, and then retrieved later to be stored in the _**DWORD**_ nShowCommand member of the shell link structure.

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: dll/win32/shell32/lang/en-US.rc (hunk 8 above)
Comment (HBelusca) [reply dll/win32/shell32/lang/en-US.rc:194]: Ah, so it is also present on "old" Windows, nice to know that. :+1:

File: dll/win32/shell32/CShellLink.cpp (hunk 10 above)
Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp:2810]: The Microsoft Shortcut specification specifically says that only those 3 SW_ values I used are allowed and I filter them when reading, therefore [it will always fit in a byte](https://github.com/reactos/reactos/pull/5638#discussion_r1312972147).

File: dll/win32/shell32/CShellLink.cpp [hunk 12]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+    for (UINT i = 0; i < _countof(runstrings); ++i)
+    {
+        WCHAR buf[MAX_PATH];
+        if (!LoadStringW(shell32_hInstance, runstrings[i], buf, _countof(buf)))
+            break;
+
+        int idx = SendMessageW(hRun, CB_ADDSTRING, 0, (LPARAM)buf);
+        if (idx < 0)
+            continue;
+        SendMessageW(hRun, CB_SETITEMDATA, idx, runshowcmd[i]);
```
Comment (katahiromz) [inline dll/win32/shell32/CShellLink.cpp:2821]:
```suggestion
        int index = SendMessageW(hRunCombo, CB_ADDSTRING, 0, (LPARAM)buf);
        if (index < 0)
            continue;
        SendMessageW(hRunCombo, CB_SETITEMDATA, index, runshowcmd[i]);
```

File: dll/win32/shell32/CShellLink.cpp [hunk 13]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+        if (!i || m_Header.nShowCommand == runshowcmd[i])
+        {
+            SendMessageW(hRun, CB_SETCURSEL, idx, 0);
```
Comment (katahiromz) [inline dll/win32/shell32/CShellLink.cpp:2824]:
```suggestion
            SendMessageW(hRunCombo, CB_SETCURSEL, index, 0);
```

File: dll/win32/shell32/CShellLink.cpp [hunk 14]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
     if (m_sDescription)
         SetDlgItemTextW(hwndDlg, IDC_SHORTCUT_COMMENT_EDIT, m_sDescription);
 
+    /* Hot key */
+    SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_KEY_HOTKEY, HKM_SETHOTKEY, m_Header.wHotKey, 0);
+
+    /* Run */
+    const WORD runstrings[] = { IDS_SHORTCUT_RUN_NORMAL, IDS_SHORTCUT_RUN_MIN, IDS_SHORTCUT_RUN_MAX };
+    const BYTE runshowcmd[] = { SW_SHOWNORMAL, SW_SHOWMINNOACTIVE, SW_SHOWMAXIMIZED };
+    HWND hRun = GetDlgItem(hwndDlg, IDC_SHORTCUT_RUN_COMBO);
```
Comment (katahiromz) [inline dll/win32/shell32/CShellLink.cpp:2811]:
```suggestion
    HWND hRunCombo = GetDlgItem(hwndDlg, IDC_SHORTCUT_RUN_COMBO);
```

File: dll/win32/shell32/CShellLink.cpp [hunk 15]
```diff
@@ -2923,6 +2946,14 @@ LRESULT CShellLink::OnNotify(HWND hwndDlg, int idFrom, LPNMHDR pnmhdr)
 
         HeapFree(GetProcessHeap(), 0, unquoted);
 
+        m_Header.wHotKey = (WORD)SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_KEY_HOTKEY, HKM_GETHOTKEY, 0, 0);
+
+        int idx = (int)SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_RUN_COMBO, CB_GETCURSEL, 0, 0);
+        if (idx != CB_ERR)
```
Comment (katahiromz) [inline dll/win32/shell32/CShellLink.cpp:2952]:
```suggestion
        int index = (int)SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_RUN_COMBO, CB_GETCURSEL, 0, 0);
        if (index != CB_ERR)
```

File: dll/win32/shell32/CShellLink.cpp [hunk 16]
```diff
@@ -2923,6 +2946,14 @@ LRESULT CShellLink::OnNotify(HWND hwndDlg, int idFrom, LPNMHDR pnmhdr)
...
+        {
+            m_Header.nShowCommand = (UINT)SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_RUN_COMBO, CB_GETITEMDATA, idx, 0);
```
Comment (katahiromz) [inline dll/win32/shell32/CShellLink.cpp:2954]:
```suggestion
            m_Header.nShowCommand = (UINT)SendDlgItemMessageW(hwndDlg, IDC_SHORTCUT_RUN_COMBO, CB_GETITEMDATA, index, 0);
```

File: dll/win32/shell32/CShellLink.cpp (hunk 14 above)
Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp:2811]: Does this really make the code better? All the related code is less than 15 lines away and it is very clear that it is a combobox because of the CB_ messages.

Comment (katahiromz) [reply dll/win32/shell32/CShellLink.cpp:2811]: Another code reader is not you.

Comment (binarymaster) [reply dll/win32/shell32/CShellLink.cpp:2811]: I agree with Katayama suggestions, going to apply them. Arguing with verified reviewers is not a good idea in any way, this only makes review process slower and delays the merge. Keep in mind, while reviewing, we are interested to help you getting the changes in.

File: dll/win32/shell32/CShellLink.cpp (hunk 10 above)
Comment (binarymaster) [reply dll/win32/shell32/CShellLink.cpp:2810]: All array members belong to enum of the type `UINT`, however MSDN documents this differently in many places, sometimes `int`, sometimes `DWORD` - all these types are 32-bit anyway. Using `BYTE` here would be an artificial limitation. What matters is not the range of values that the specification dictates, it's the internals, i.e. the way how it's stored in the shortcut. Imagine one day someone would want to extend the range without breaking the file format...

File: dll/win32/shell32/CShellLink.cpp [hunk 17]
```diff
@@ -2802,6 +2802,29 @@ BOOL CShellLink::OnInitDialog(HWND hwndDlg, HWND hwndFocus, LPARAM lParam)
...
+            break;
+
+        int index = SendMessageW(hRunCombo, CB_ADDSTRING, 0, (LPARAM)buf);
+        if (index < 0)
+            continue;
+        SendMessageW(hRunCombo, CB_SETITEMDATA, index, runshowcmd[i]);
+        if (!i || m_Header.nShowCommand == runshowcmd[i])
+        {
+            SendMessageW(hRunCombo, CB_SETCURSEL, index, 0);
+        }
```
Comment (binarymaster) [inline dll/win32/shell32/CShellLink.cpp:2825]: Style consistency with two other if's above:
```suggestion
            SendMessageW(hRunCombo, CB_SETCURSEL, index, 0);
```

**Review by HBelusca — APPROVED** [review]

File: dll/win32/shell32/CShellLink.cpp (hunk 10 above)
Comment (whindsaks) [reply dll/win32/shell32/CShellLink.cpp]: > Using `BYTE` here would be an artificial limitation.

There is no BYTE limitation. I am dictating which values are stored in the array. Clearly this is a hardcoded list that will never overflow a BYTE. If Microsoft somehow decides to add new SW_ values that are larger than a BYTE **and** change the .lnk spec to allow them, then yes, it would have to change to a DWORD.

**Review by tkreuzer — APPROVED** [review]
